### DISCUSSION TITLE
How long to wait for an API callback other than SendData

### DISCUSSION THREAD

**Discussion by AlCalzone** [zwave-js/zwave-js#1786] [General] [answered]
@hanskroner it's me again :)

This time I'm wondering which timeouts (if any) to apply to Serial API callbacks other than SendData. I don't remember having anything in the specs, except the 65 second timeout of SendData. I'm currently applying it to all API calls which seemed to have gone well, except for this case:

https://github.com/zwave-js/node-zwave-js/issues/1764

Here, the controller requests Node 2 to update its neighbors and quickly gets a callback that the process was started. For the next 65 seconds, nothing happens and the API call times out. This continues with the next node and the next ...

I'm not sure if we should even wait for the Completed Callback or just continue sending messages. The specs don't seem to give any advice regarding the second callback.

**Comment by hanskroner** [comment, marked as answer]
> I don't remember having anything in the specs, except the 65 second timeout of SendData.

If memory serves, those 65 seconds are an estimate built up from the maximum allowed frame length, the maximum number of repeater hops - with the last one beaming to a FLiRS node - and the maximum number of retries, with some estimation for the maximum random back-off and LBT the protocol would do in a very busy network.

> Here, the controller requests Node 2 to update its neighbors and quickly gets a callback that the process was started. For the next 65 seconds, nothing happens and the API call times out.

To get this timeout right, you need to understand what the node is doing when you ask it to look for neighbors. Along with the command, the node receives a bitmask containing all the always-listening nodes (including FLiRS) the controller knows about. The controller is asking the node to try to reach all of those devices and report the results back. This is the same procedure the controller asks the a node to perform when it has just been included into a network, and the reason why the timeout for `ZW_AddNodeToNetwork` is variable and can be pretty big - it depends on the network size and the types of devices that make it up. You should follow the timeout recommendations for `ZW_AddNodeToNetwork` when asking a node to look for neighbors:

For slave nodes:
```
7600ms + (NUMBER_OF_LISTENING_NODES * 217ms) + (NUMBER_OF_FLIRS_NODES * 3517ms)
```

For controller nodes:
```
7600ms + (NUMBER_OF_LISTENING_NODES * 217ms) + (NUMBER_OF_FLIRS_NODES * 3517ms) + (TOTAL_NUMBER_OF_NODES * 732ms)
```

"Network Healing" sounds harmless, but in reality is a process that typically takes hours even on small/medium sized networks. It wouldn't hurt to let the library's clients about this, so they can show appropriate warnings to users attempting to trigger this. While the Z-Wave network doesn't actually go "down" during this process, there is a lot of traffic generated and very reduced performance is to be expected while the operation is running.

As for what's going on in the issue you linked, it'll be very hard to tell without traces of the actual communication. I can think of two things: 1) The network in question has a size/device makeup for which 65s is not enough time for a node to ping all listening devices. 2) You're asking a sleeping device to look for neighbors, and it's falling asleep in the process, never replying - this would be very odd, but I've heard of it happening before. 

To my knowledge, the SerialAPI command for asking a node to look for neighbors hasn't changed from the 500-series hardware to Z-Wave 700.

> I'm not sure if we should even wait for the Completed Callback or just continue sending messages. The specs don't seem to give any advice regarding the second callback.

You definitely should wait for the second callback (or the timeout) if the process was started. The docs mention you should expect a second `REQ` callback if the first one returns `REQUEST_NEIGHBOR_UPDATE_STARTED`.

**Reply by AlCalzone** [reply]
> 2) You're asking a sleeping device to look for neighbors, and it's falling asleep in the process, never replying

Since the problem affects every single node, I would guess that this is not the case.

> You definitely should wait for the second callback (or the timeout) if the process was started. The docs mention you should expect a second `REQ` callback if the first one returns `REQUEST_NEIGHBOR_UPDATE_STARTED`.

Okay, we currently do that (except that we assume that the callback timed out after 65s). What also happens is that the send queue is blocked for that time, so no other messages except nonce replies go out. This is terrible from the users perspective, so I'm wondering if waiting for the 2nd callback could happen in the background.

**Reply by hanskroner** [reply]
>This is terrible from the users perspective, so I'm wondering if waiting for the 2nd callback could happen in the background.

It shouldn't be. Like I mentioned above, it should be made clear to the user that this is not an operation that should be done when the network is expected to operate at full performance. The controller should not be sending out commands during this time. This operation really needs to be scheduled when nothing else is going on - it's not an operation that can be though of as "running in the background."

**Reply by AlCalzone** [reply]
Seems like it was a connectivity issue and/or firmware issue after all:
https://github.com/zwave-js/node-zwave-js/issues/1764#issuecomment-780865266

**Reply by hanskroner** [reply]
> Seems like it was a connectivity issue and/or firmware issue after all:
> [#1764 (comment)](https://github.com/zwave-js/node-zwave-js/issues/1764#issuecomment-780865266)

No, I really don't think that was it - the changelogs don't show anything even remotely relevant.

The report seems to come from the US, and it "seems" that bumping the output power made things work for him. First off, I'm not sure where the legal line is on these things, but in the US the FCC restricts output on the ISM bands Z-Wave uses to +0dBm - he's not allowed to go above that and recommending anyone in the US to change this setting is a bad idea. For us in EU, the limit is +13.5dBm.

It might be that this controller is in a very RF-noisy environment, with regards to the 916MHz bands Z-Wave uses in the US. The 500-series had a physical SAW filter to help block out interference. Z-Wave 700 has better blocking performance than the 500-series, and typically doesn't need the a SAW filter, so chances are the USB stick he's using doesn't have one - but in this particular case might be affected by not having it. Since there isn't a sniffer available, the next best thing would be to use the additional data SendData gives to look at the RSSI values for messages to the failing devices - if they manage to reply. I would also ask that he tries to re-position his controller away from interference sources, keeping the power output at the legal max. of +0dBm.
